### PR TITLE
389 combobox fixes

### LINKED ISSUE #369: Combobox: Selected option in scrollable options list needs to remain visible
# Bug Report

## Description
Skin fixed an issue on the combobox where the size of the item list is now limited, and introduces a scroll.   However, if a keyboard user starts navigating down the list and gets beyond the visible area then the selected options aren't visible.  

We need to keep the scroll of the options container in sync with the selected item.

This is also an issue if the selected item is out of view when the dropdown is first opened, even using the mouse.

**We'll need this same fix for menu and the select using listbox**

## Workaround
None

## Screenshots
screen shot 2018-08-24 at 9 38 37 am

### PR DESCRIPTION
**Pull request by seangates**
## Description
- add a scroll utility for element scrolling

## Context
When opening a menu or combobox, or navigating using the keyboard within those components, the element chosen should move the element into view.

## References
Fixes #369.

### CODE REVIEW

File: src/components/ebay-menu/index.js [hunk 1]
```diff
@@ -12,6 +13,7 @@ const mainButtonClass = 'expand-btn';
 const buttonSelector = `.${mainButtonClass}`;
 const contentClass = 'expander__content';
 const contentSelector = `.${contentClass}`;
+const checkedItemSelector = '.menu__item[role=menuitemradio][aria-checked=true]';
```
[the remark this answers is not in the document]
Comment (seangates) [reply src/components/ebay-menu/index.js]: Oh, let me fix the selector.

File: src/components/ebay-menu/examples/14-scrollable/template.marko [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<var i=0/>
+<ebay-menu text="eBay Menu" type="radio">
+    <ebay-menu-item for(idx from 1 to 100) checked=(idx === 24)>
```
[the remark this answers is not in the document]
Comment (seangates) [reply src/components/ebay-menu/examples/14-scrollable/template.marko]: Marko v3, ~FTW~.

File: src/components/ebay-combobox/examples/07-scrollable/template.marko [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<ebay-combobox name="formFieldName">
+    <for(i from 1 to 100)>
+        <ebay-combobox-option value=i text=`Option ${i}` selected=(i === 25)/>
```
[the remark this answers is not in the document]
Comment (seangates) [reply src/components/ebay-combobox/examples/07-scrollable/template.marko]: Good catch. I'll update.

File: src/common/element-scroll/index.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+module.exports = { scroll, nodeListToArray };
```
[the remark this answers is not in the document]
Comment (seangates) [reply src/common/element-scroll/index.js]: https://github.com/eBay/ebayui-core/blob/master/src/common/body-scroll/index.js#L6

I'll change it, though.

File: src/common/element-scroll/test/test.browser.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        const innerDiv = `<div class="item${index}" style="height: 50px">content</div>`;
+        contentDiv.innerHTML += innerDiv;
+    }
+
+    const innerEl1 = elementScroll.nodeListToArray(contentDiv.querySelectorAll('.item5'))[0];
+    const innerEl2 = elementScroll.nodeListToArray(contentDiv.querySelectorAll('.item2'))[0];
+
+    before(() => {
+        document.body.appendChild(contentDiv);
+        contentDiv.setAttribute('style', 'max-height:100px; overflow-y: auto;');
```
[the remark this answers is not in the document]
Comment (seangates) [reply src/common/element-scroll/test/test.browser.js]: Didn't work. It was somewhow `undefined` when set there.

File: src/common/element-scroll/test/test.browser.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+describe('element-scroll', () => {
+    const contentDiv = document.createElement('div');
+
+    for (let index = 0; index < 10; index += 1) {
+        const innerDiv = `<div class="item${index}" style="height: 50px">content</div>`;
+        contentDiv.innerHTML += innerDiv;
+    }
+
+    const innerEl1 = elementScroll.nodeListToArray(contentDiv.querySelectorAll('.item5'))[0];
+    const innerEl2 = elementScroll.nodeListToArray(contentDiv.querySelectorAll('.item2'))[0];
```
[the remark this answers is not in the document]
Comment (seangates) [reply src/common/element-scroll/test/test.browser.js]: 👍

File: src/common/element-scroll/index.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    const scrollBottom = parentEl.scrollTop + parentEl.offsetHeight;
+
+    if (el.offsetTop < parentEl.scrollTop) {
+        parentEl.scrollTop = el.offsetTop;
+    } else if (offsetBottom > scrollBottom) {
+        parentEl.scrollTop = offsetBottom - parentEl.offsetHeight;
+    }
+}
+
+function nodeListToArray(nodeList) {
```
[the remark this answers is not in the document]
Comment (seangates) [reply src/common/element-scroll/index.js]: 👍

File: src/components/ebay-menu/index.js [hunk 8]
```diff
@@ -213,6 +215,8 @@ function processAfterStateChange(itemIndexes) {
         });
     }
 
+    elementScroll.scroll(itemEl);
```
Comment (yomed) [inline src/components/ebay-menu/index.js]: I think we should probably wrap this (and the one below) in `if (this.state.isRadio || this.state.isCheckbox)` so we avoid querying the DOM unnecessarily.

File: src/components/ebay-menu/index.js [hunk 9]
```diff
@@ -290,6 +294,8 @@ function handleButtonEscape() {
 }
 
 function handleExpand() {
+    const selectedOptionEl = this.el.querySelectorAll(checkedItemSelector)[0];
+    elementScroll.scroll(selectedOptionEl);
```
Comment (yomed) [inline src/components/ebay-menu/index.js]: `elementScroll.scroll(this.el.querySelector(checkedItemSelector));`

File: src/common/element-scroll/test/test.browser.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+describe('element-scroll', () => {
+    const contentDiv = document.createElement('div');
+
+    for (let index = 0; index < 10; index += 1) {
+        const innerDiv = `<div class="item${index}" style="height: 50px">content</div>`;
+        contentDiv.innerHTML += innerDiv;
+    }
+
+    const innerEl1 = contentDiv.querySelectorAll('.item5')[0];
+    const innerEl2 = contentDiv.querySelectorAll('.item2')[0];
```
Comment (yomed) [inline src/common/element-scroll/test/test.browser.js]: I'd suggest renaming these like `secondItemEl fifthItemEl`
Can also use `querySelector()` in place of `querySelectorAll()[0]`

File: src/components/ebay-combobox/index.js [hunk 11]
```diff
@@ -101,6 +103,8 @@ function init() {
 }
 
 function handleExpand() {
+    const selectedOptionEl = this.el.querySelectorAll(comboboxSelectedOptionSelector)[0];
```
Comment (yomed) [inline src/components/ebay-combobox/index.js]: `querySelector()` in place of `querySelectorAll()[0]`

**Comment by msendlakowski** [comment]
Will this move the scroll for a keyboard user who is navigating the list?  I thought we'd have to add a listener to Ian's rovingTabIndex plugin.  
`rovingTabindexChange` event here:
https://github.com/makeup-js/makeup-roving-tabindex#example

**Comment by seangates** [comment]
@msendlakowski It does move, but only when a selection is made. Also, the Combobox does not use the roving-tab-index, but the menu does. This change works for both whenever there is processing after the a selection is made.

Have you tried it out locally yet? You'll see what I mean.

**Comment by seangates** [comment]
@yomed Updated once more. Please review again.

**Comment by msendlakowski** [comment]
I'm going to merge this so I can get a new pre-release out.  @yomed  if you have any remaining issues @seangates can follow up

**Pull request merged**